packages/plugin-js-packages/src/lib/utils.unit.test.ts

```diff
@@ -0,0 +1,301 @@
+    await expect(derivePackageManagerInPackageJson()).resolves.toBe(
+      'yarn-classic',
+    );
+    expect(fileExistsSpy).toHaveBeenCalledWith('/test/package.json');
+    expect(exec).toHaveBeenCalledTimes(0);
+  });
+
+  it('should return yarn-modern if packageManager field in package.json is yarn v2', async () => {
+    vol.fromJSON(
+      {
+        'package.json': JSON.stringify({
+          packageManager:
+            'yarn@2.2.3+sha224.953c8233f7a92884eee2de69a1b92d1f2ec1655e66d08071ba9a02fa',
+        }),
+      },
+      MEMFS_VOLUME,
+    );
+
+    await expect(derivePackageManagerInPackageJson()).resolves.toBe(
+      'yarn-modern',
+    );
+    expect(fileExistsSpy).toHaveBeenCalledWith('/test/package.json');
+    expect(exec).toHaveBeenCalledTimes(0);
+  });
+
+  it('should return yarn-modern if packageManager field in package.json is yarn v3', async () => {
+    vol.fromJSON(
+      {
+        'package.json': JSON.stringify({
+          packageManager:
+            'yarn@3.2.3+sha224.953c8233f7a92884eee2de69a1b92d1f2ec1655e66d08071ba9a02fa',
+        }),
+      },
+      MEMFS_VOLUME,
+    );
+
+    await expect(derivePackageManagerInPackageJson()).resolves.toBe(
+      'yarn-modern',
+    );
+    expect(fileExistsSpy).toHaveBeenCalledWith('/test/package.json');
+    expect(exec).toHaveBeenCalledTimes(0);
+  });
+});
+
+describe('derivePackageManager', () => {
+  const fileExistsSpy = vi.spyOn(utils, 'fileExists');
+
+  beforeEach(() => {
+    fileExistsSpy.mockClear();
+  });
+
+  it('should return packageManager from field in package.json', async () => {
+    vol.fromJSON(
+      {
+        'package.json': JSON.stringify({
+          packageManager:
+            'pnpm@3.2.3+sha224.953c8233f7a92884eee2de69a1b92d1f2ec1655e66d08071ba9a02fa',
+        }),
+      },
+      MEMFS_VOLUME,
+    );
+
+    await expect(derivePackageManager()).resolves.toBe('pnpm');
+    expect(fileExistsSpy).toHaveBeenCalledWith('/test/package.json');
+    expect(exec).toHaveBeenCalledTimes(0);
+  });
+
+  it('should return npm if a package.jock.json is present', async () => {
+    vol.fromJSON(
+      {
+        'package-lock.json': '{}',
+      },
+      MEMFS_VOLUME,
+    );
+    await expect(derivePackageManager()).resolves.toBe('npm');
+    expect(fileExistsSpy).toHaveBeenCalledWith('/test/package-lock.json');
+    expect(exec).not.toHaveBeenCalled();
+  });
+
+  it('should return pnpm if pnpm-lock.yaml is present', async () => {
+    vol.fromJSON(
+      {
+        'pnpm-lock.yaml': '',
+      },
+      MEMFS_VOLUME,
+    );
+    await expect(derivePackageManager()).resolves.toBe('pnpm');
+    expect(fileExistsSpy).toHaveBeenCalledWith('/test/pnpm-lock.yaml');
+    expect(exec).not.toHaveBeenCalled();
+  });
+
+  it('should return yarn-classic if yarn.lock is present', async () => {
+    vol.fromJSON(
+      {
+        'yarn.lock': '',
+      },
+      MEMFS_VOLUME,
+    );
+    (exec as MockInstance<[], unknown>).mockImplementation((_, fn) =>
```

Comment:
Could need some help to type exec mock :)

Reply:
I'd suggest the following approach:

```ts
import cp from 'node:child_process';

// ...

describe('...', () => {
  // 👇 casting narrows mockImplementation types
  const execSpy = vi.spyOn(cp, 'exec') as MockInstance<
    [string, (error: Error | null, stdout: string, stderr?: string) => void]
  >;

  afterAll(() => {
    execSpy.mockRestore();
  });

  it('...', () => {
    // ...
    execSpy.mockImplementation((_, callback) => {
      callback(null, '1.22.19');
    });
    // ...
    expect(execSpy).toHaveBeenCalledWith('yarn -v', expect.any(Function));
  });
});
```

Reply:
I can't define the spy at the top of the describe with spyOn. It throws: 
`TypeError: Cannot redefine property: exec`. I don't know where but we had that problem already somewhere with `child_process`.

Reply:
I refactored it and pushed so someone could reproduce the case.

Reply:
Ah, it's possible this spying doesn't work for built-in `node:` modules.

On seconds thoughts, why not just use our `executeProcess` function from `utils`? It's a lot easier to mock as it's already promisifed.

![image](https://github.com/user-attachments/assets/43c23cee-0e25-4bac-bdce-d1c2d0dc6d35)

Reply:
what I learned:
- All node internal seems to work only with `vi.mock('import', () => { })`
- `vi-spyOn(obj, prop)` only works on the shallow level of a function. Meaning if I spy on A and test C which uses B which intern uses A (C->B->A). Means B is using the original function